integrations/gitlab_v2/gitlab_integration.py

```diff
@@ -0,0 +1,185 @@
+    "open",
+    "reopen",
+    "update",
+    "approved",
+    "unapproved",
+    "approval",
+    "unapproval",
+]
+DELETE_WEBHOOK_EVENTS: list[str] = ["close", "merge"]
+
+class GitLabIntegration:
+    def __init__(self):
+        self.gitlab_handlers = []
+        self.webhook_handlers = []
+        self.event_handlers = self._register_event_handlers()
+
+    @staticmethod
+    def _validate_configuration():
+        if not ocean.integration_config["gitlab_access_tokens"]:
+            raise ValueError("No GitLab Tokens provided in configuration")
+        if not ocean.integration_config.get("gitlab_host"):
+            raise ValueError("GitLab host not provided in configuration")
+
+    @staticmethod
+    def _get_gitlab_tokens():
+        gitlab_tokens = ocean.integration_config["gitlab_access_tokens"]
+        if isinstance(gitlab_tokens, str):
+            gitlab_tokens = [gitlab_tokens]
+        return [token for token in gitlab_tokens if token]
+
+    @staticmethod
+    def _extract_id_from_payload(payload: Dict[str, Any], key: str) -> str:
+        return payload.get(key, {}).get("id")
+
+    @staticmethod
+    def _determine_ocean_action(object_attributes_action: str) -> Any | None:
+        if object_attributes_action in DELETE_WEBHOOK_EVENTS:
+            return ocean.unregister_raw
+        elif object_attributes_action in CREATE_UPDATE_WEBHOOK_EVENTS:
+            return ocean.register_raw
+        return None
+
+    async def initialize(self):
+        self._validate_configuration()
+
+        gitlab_tokens = self._get_gitlab_tokens()
+        gitlab_host = ocean.integration_config["gitlab_host"]
+
+        logger.info(f"Initializing with {len(gitlab_tokens)} tokens")
+
+        for token in gitlab_tokens:
+            gitlab_handler = GitlabClient(gitlab_host, token)
+            webhook_handler = WebhookHandler(gitlab_handler)
+            self.gitlab_handlers.append(gitlab_handler)
+            self.webhook_handlers.append(webhook_handler)
+
+        logger.info(f"GitLab integration initialized with {len(self.gitlab_handlers)} handlers")
+
+        await self.setup_webhooks()
+
+    async def setup_webhooks(self):
+        logger.info(f"Setting up webhooks with events: {SUPPORTED_EVENTS}")
+
+        setup_tasks = [
+            asyncio.create_task(handler.setup_group_webhooks(WEBHOOK_URL, SUPPORTED_EVENTS))
+            for handler in self.webhook_handlers
+        ]
+
+        try:
+            await asyncio.gather(*setup_tasks)
+            logger.info("Webhooks set up successfully for all GitLab instances")
+        except Exception as e:
+            logger.error(f"Failed to set up webhooks: {str(e)}")
+            raise
+
+    async def resync_resources(self, kind: ResourceKind, kind_configs: Any) -> AsyncGenerator[Dict[str, Any], None]:
+        endpoint = RESOURCE_ENDPOINT_MAPPING.get(kind)
+        if not endpoint:
+            raise ValueError(f"Invalid ObjectKind: {kind}")
```

Comment:
actually here we would love if you can still query the _fetch_resources() so that the code can still work for kind such as pipeline, users, jobs etc. this way, the integration will be generic enough to attend to kinds that have not been explicitly defined 

this way, once you get a kind such as pipelines, the code will make call to https://gitlab.com/api/v4/pipeline, which is similar to how you query projects, issues, merge requests etc

Reply:
Thank you for the feedback! The `_fetch_resources` method is already designed to be generic and can handle any resource kind by taking the dynamically constructed endpoint.

When a resource kind like pipeline, jobs, or users is provided, it constructs the appropriate endpoint (e.g., pipelines, jobs, users) and queries it via the same logic used for querying projects, issues, or merge requests. This way, the integration will work for both explicitly defined kinds and those that are not mapped, as long as a valid endpoint can be dynamically constructed.

Let me know if you'd like any further clarification or adjustments!